flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveSourceDynamicFileEnumerator.java

```diff
@@ -0,0 +1,218 @@
+        this.table = checkNotNull(table);
+        this.dynamicPartitionKeys = checkNotNull(dynamicPartitionKeys);
+        this.allPartitions = checkNotNull(allPartitions);
+        this.threadNum = threadNum;
+        this.jobConf = checkNotNull(jobConf);
+
+        this.finalPartitions = this.allPartitions;
+    }
+
+    public void setDynamicFilteringData(DynamicFilteringData data) {
+        LOG.debug("Filtering partitions of table {} based on the data: {}", table, data);
+        if (!data.isFiltering()) {
+            finalPartitions = allPartitions;
+            return;
+        }
+        finalPartitions = new ArrayList<>();
+        RowType rowType = data.getRowType();
+        Preconditions.checkArgument(rowType.getFieldCount() == dynamicPartitionKeys.size());
+        for (HiveTablePartition partition : allPartitions) {
+            RowData partitionRow = createRowData(rowType, partition.getPartitionSpec());
+            if (data.contains(partitionRow)) {
+                finalPartitions.add(partition);
+            }
+        }
+        LOG.info(
+                "Dynamic filtering table {}, original partition number is {}, remaining partition number {}",
+                table,
+                allPartitions.size(),
+                finalPartitions.size());
+    }
+
+    @VisibleForTesting
+    RowData createRowData(RowType rowType, Map<String, String> partitionSpec) {
+        GenericRowData rowData = new GenericRowData(rowType.getFieldCount());
+        for (int i = 0; i < rowType.getFieldCount(); ++i) {
+            String value = partitionSpec.get(dynamicPartitionKeys.get(i));
+            Object convertedValue;
+            if (JobConfUtils.getDefaultPartitionName(jobConf).equals(value)) {
+                // Keep the same way as hive source reads from default partition. Use the default
+                // partition name if the partition field is string type, use null otherwise.
+                // See HivePartitionUtils#restorePartitionValueFromType.
+
+                // By default, the type of hive partition field is nullable.
+                // See HiveTypeUtil#toFlinkPrimitiveType
+                if (rowType.getTypeAt(i).is(LogicalTypeFamily.CHARACTER_STRING)) {
+                    convertedValue = StringData.fromString(value);
+                } else {
+                    convertedValue = null;
+                }
+            } else {
+                switch (rowType.getTypeAt(i).getTypeRoot()) {
+                    case CHAR:
+                    case VARCHAR:
+                        convertedValue = StringData.fromString(value);
+                        break;
+                    case TINYINT:
+                        convertedValue = Byte.parseByte(value);
+                        break;
+                    case SMALLINT:
+                        convertedValue = Short.parseShort(value);
+                        break;
+                    case INTEGER:
+                    case DATE:
```

Comment:
Since when the partition's type `DATE`, the value will be `yyyy-[m]m-[d]d`, we can't convert the value to `Integer` directly.
We can follow the convert logic in `HivePartitionUtils#restorePartitionValueFromType`

Reply:
Thanks for the correction! I'm wondering if we can parse all types with the util method?